improv_ble/improv_ble_client.py

```diff
@@ -0,0 +1,469 @@
+                raise
+
+    async def _ensure_connected(self) -> None:
+        """Ensure connection to device is established."""
+        if self._connect_lock.locked():
+            _LOGGER.debug(
+                "%s: Connection already in progress, waiting for it to complete; "
+                "RSSI: %s",
+                self.name,
+                self.rssi,
+            )
+        if self._client and self._client.is_connected:
+            self._reset_disconnect_timer()
+            return
+        async with self._connect_lock:
+            # Check again while holding the lock
+            if self._client and self._client.is_connected:
+                self._reset_disconnect_timer()
+                return
+            _LOGGER.debug("%s: Connecting; RSSI: %s", self.name, self.rssi)
+            client = await establish_connection(
+                BleakClientWithServiceCache,
+                self._ble_device,
+                self.name,
+                self._disconnected,
+                use_services_cache=False,  # True
+                ble_device_callback=lambda: self._ble_device,
+            )
+            _LOGGER.debug("%s: Connected; RSSI: %s", self.name, self.rssi)
+
+            self._client = client
+            self._disconnect_reason = None
+            self._reset_disconnect_timer()
+
+            _LOGGER.debug(
+                "%s: Subscribe to notifications; RSSI: %s", self.name, self.rssi
+            )
+            await client.start_notify(
+                CHARACTERISTIC_UUID_ERROR, self._notification_handler
+            )
+            await client.start_notify(
+                CHARACTERISTIC_UUID_RPC_RESULT, self._rpc_result_handler
+            )
+            await client.start_notify(
+                CHARACTERISTIC_UUID_STATE, self._notification_handler
+            )
+
+    def _raise_if_not_connected(self) -> None:
+        """Raise if the connection to device is lost."""
+        if self._client and self._client.is_connected:
+            self._reset_disconnect_timer()
+            return
+        raise NotConnected
+
+    def _reset_disconnect_timer(self) -> None:
+        """Reset disconnect timer.
+
+        If the disconnect timer expires, disconnect from the device.
+        """
+
+        async def _disconnect() -> None:
+            """Execute disconnect request."""
+            _LOGGER.debug(
+                "%s: Disconnecting after timeout of %s",
+                self.name,
+                DISCONNECT_DELAY,
+            )
+            await self._execute_disconnect(DisconnectReason.TIMEOUT)
+
+        def _schedule_disconnect() -> None:
+            self._disconnect_timer = None
+            asyncio.create_task(_disconnect())
+
+        if self._disconnect_timer:
+            self._disconnect_timer.cancel()
+        self._expected_disconnect = False
+        self._disconnect_timer = self.loop.call_later(
+            DISCONNECT_DELAY, _schedule_disconnect
```

Comment:
Alternatively we could sleep in the coroutine function before actually disconnecting. Then we don't need to use the event loop explicitly here `loop.call_later`.

Reply:
We want to to be able to interrupt the sleeping though, what pattern do you suggest for that?

Reply:
I haven't looked at the code, but if its awaiting a future 

https://github.com/bdraco/async_interrupt

Reply:
I think we can just cancel the task.

Reply:
But since we need the loop to create futures anyway, we don't gain much by avoiding the loop for this case. I generally try to avoid using the loop directly as far as possible in asyncio projects but sometimes it's hard to avoid.